Add custom runtime documentation (#140)
rancher/k3k-product-docs: versions/v1.1.0/modules/en/pages/how-tos-for-admin/custom-runtime-class.adoc

= Using a Custom Container Runtime
:revdate: 2026-04-28
:page-revdate: {revdate}

> By default, {k3k-product-name} runs virtual cluster server and agent pods using the host cluster's default container runtime (typically `runc`). You can override this by setting the `runtimeClassName` field on the `Cluster` resource, which lets virtual clusters run on alternative runtimes such as https://github.com/containers/crun[crun], https://katacontainers.io/[Kata Containers], or https://gvisor.dev/[gVisor].

Reviewer: ```suggestion
By default, {k3k-product-name} runs virtual cluster server and agent pods using the host cluster's default container runtime (typically `runc`). You can override this by setting the `runtimeClassName` field on the `Cluster` resource, which lets virtual clusters run on alternative runtimes, such as https://github.com/containers/crun[crun], https://katacontainers.io/[Kata Containers], or https://gvisor.dev/[gVisor].
```

> This guide walks through configuring a host cluster with `crun` and creating a virtual cluster that uses it.

Reviewer: ```suggestion
This guide walks you through configuring a host cluster with `crun` and creating a virtual cluster that uses it.
```

== Prepare the host cluster

> The runtime must be installed and registered with containerd on the host cluster before it can be referenced from a virtual cluster.

Reviewer: ```suggestion
Follow the below steps to install and register the runtime with `containerd` on the host cluster before referencing it from a virtual cluster:
```

> . Install K3s on the host, if it is not already running. See the https://docs.k3s.io/[K3s documentation] for installation instructions.

Reviewer: ```suggestion
. Install K3s on the host, if it is not already running. For installation instructions, see https://docs.k3s.io/[K3s documentation].
```

. Download the `crun` binary from the https://github.com/containers/crun/releases[crun releases page] and place it on the host:
+
[source,sh]
----
mv crun.amd64 /usr/local/bin/crun
chmod +x /usr/local/bin/crun
----

> . Restart K3s so that containerd picks up the new runtime:

Reviewer: ```suggestion
. Restart K3s so that `containerd` picks up the new runtime:
```

+
[source,sh]
----
systemctl restart k3s
----

. Confirm that K3s auto-detected `crun`. The following stanza should appear in `/var/lib/rancher/k3s/agent/etc/containerd/config.toml`:
+
[source,toml]
----
[plugins."io.containerd.grpc.v1.cri".containerd.runtimes."crun"]
  runtime_type = "io.containerd.runc.v2"
[plugins."io.containerd.grpc.v1.cri".containerd.runtimes."crun".options]
  BinaryName = "/usr/local/bin/crun"
  SystemdCgroup = true
----
+

> K3s also creates a matching `RuntimeClass` resource in the cluster, which is what the `runtimeClassName` field on the virtual cluster will reference.

Reviewer: This can be a verification section as it isn't a strictly procedural step.

```suggestion
=== Verification

. Confirm whether K3s auto-detected `crun`by verifying whether the following appears in `/var/lib/rancher/k3s/agent/etc/containerd/config.toml`:
+
[source,toml]
----
[plugins."io.containerd.grpc.v1.cri".containerd.runtimes."crun"]
  runtime_type = "io.containerd.runc.v2"
[plugins."io.containerd.grpc.v1.cri".containerd.runtimes."crun".options]
  BinaryName = "/usr/local/bin/crun"
  SystemdCgroup = true
----
+
K3s also creates a matching `RuntimeClass` resource in the cluster, which is what the `runtimeClassName` field on the virtual cluster will reference.
```
